File: protect/src/main/java/com/webileapps/safeguard/NetworkMonitor.java
package com.webileapps.safeguard;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;

public class NetworkMonitor {
    private final Context context;
    private final ConnectivityManager connectivityManager;
    private NetworkChangeCallback networkCallback;

    public NetworkMonitor(Context context) {
        this.context = context;
        this.connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    public void startMonitoring(NetworkChangeCallback.NetworkChangeListener onNetworkChange) {
        networkCallback = new NetworkChangeCallback(onNetworkChange);
        NetworkRequest networkRequest = new NetworkRequest.Builder()
                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                .build();
        connectivityManager.registerNetworkCallback(networkRequest, networkCallback);
    }

    public void stopMonitoring() {
        if (networkCallback != null) {
            connectivityManager.unregisterNetworkCallback(networkCallback);
        }
    }
}

File: protect/src/main/java/com/kfintech/protect/NetworkMonitor.kt
package com.kfintech.protect

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities

class NetworkMonitor(private val context: Context) {
    private val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private var networkCallback: NetworkChangeCallback? = null

    fun startMonitoring(onNetworkChange: (Boolean) -> Unit) {
        networkCallback = NetworkChangeCallback(onNetworkChange)
        val networkRequest = android.net.NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        connectivityManager.registerNetworkCallback(networkRequest, networkCallback!!)
    }

    fun stopMonitoring() {
        networkCallback?.let {
            connectivityManager.unregisterNetworkCallback(it)
        }
    }
}